# NumPy kutubxonasi

## Kirish

Har bir sohada bo'lgani kabi, biz ushbu sohada ham imkon qadar mavjud narsalardan unimli foydalanishga harakat qilishimiz zarur bo'ladi. Chunki ular yillar davomida sinovdan o'tgan hamda ma'qullangan bo'ladi. Shuning uchun ham ushbu bo'limda biz SI uchun eng zarur bo'lgan kutubxona -- NumPyni qisman o'rganib chiqamiz.

NumPy kutbxonasi Python dasturlash tilida bir turdagi n o'lchamli tensorlarni saqlash uchun hamda ular ustida algebraik amallarni bajarish uchun zarur hisoblanadi. U 2005 yilda birinchi bo'lib "Numeric" va "Numarray" kutubxonalarining davomchisi sifatida ishlab chiqila boshladi. Ushbu kutubxona Python dasturlash tilida hamda to'lig'icha ochiq kodli bo'lib, u sonli hisoblar uchun mo'ljallangan.

> Biz odatda vektor deb bir o'lchamli, matritsa deb esa 2 o'lchamli tensorga aytamiz, ya'ni (xatolik bilan aytganda) vektor va matritsaning umumiyroq ko'rinishi tensor deyiladi. Vektorlar, odatda, qator vektor yoki ustun vektor bo'lishi mumkin. Xuddi shu qator-ustun tushunchasini biz matritsalarga ham ishlatamiz. Lekin 3 o'lchovli tensorlardan boshlab bunday tushunchalarni kiritishning iloji yo'q, ya'ni nom bera olmaymiz. Shuning uchun ham odatda biz har bir o'lchovli "O'Q" deb ataymiz. Masalan, matritsa qatori birinchi o'q bo'lsa, ustini ikkinchi o'q bo'ladi. Umumiy holda esa, n-o'q deb ataymiz n-o'lchovli tensorlarda.


Biz ushbu kutubxonani Python dasturlash tilida o'rgangan `list` hamda `tuple` turlarining o'rniga ishlatamiz. Bu esa bizga o'z navbatida bir qancha ustunliklarni taqdim etadi:

1. Tezlik. C tilida maxsus kuchaytirgichlar bilan yozilganligi uchun u nafaqat Pythonda yozilgan sodda kodlardan balki C/C++ tilida yozilgan kodlardan ham ancha samarali ishlaydi.
2. Funksiyalari. Chiziqli algebra hamda matritsalar ustida mavjud bo'lgan amallarning ko'pchiligi mavjudligi.
3. Asosiy kutubxona. Bir qancha boshqa kutubxonalar uchun asos tur sifatida xizmat qiladi.
4. Qolipi. Ushbu kutubxonaning yondashuvi asosida hozirgi zamonaviy chuqur o'rganish kutubxonalari yozilgan. Ya'ni bu kutubxonani o'rganish bilan PyTorch hamda Tensorflow kutubxonalarining tensorlari bilan ishlashni ham o'rganish mumkin.
5. Soddalik, qisqalik va tushanarlilik. Ushbu kutubxonadan foydalanish juda sodda bo'lib, biz bir qancha qatorlarda yozadigan kodni bir necha qatorda tushanarli qilib yozish imkoniyati mavjud.

Biz ushbu bo'limda Numpyning asoslarini o'rganib chiqamiz, keyinchalik esa ushbu sinfning boshqa yuqori imkoniyatlari bilan yana tanishib chiqamiz.

## Numpy o'rnatish va qo'shish

Ushbu kutubxonani biz `pip install numpy` buyrug'i orqali osongina o'rnatib olishimiz mumkin. Asosiy kodga qo'shishda esa uni `np` deb qisqartirgan holda amalga oshiramiz:

In [1]:
import numpy as np

## Obyekt yaratish

Obyektlarni bir qancha usullar bilan yaratish mumkin, ular ichida biror boshqa turdagi o'zgaruvchidan ishbu turga o'tkazish ham fayldan o'qish ham mavjud. Keling, avvalo oddiy Python `list` obyektini ushbu turga o'zkazib ushbu kutubxonaning asoslarini o'rganaylik, keyin esa fayldan o'qishni ko'rib chiqamiz.

Buning uchun biz kutobxonaning `array()` funksiyasidan chaqirishib, uning birinchi argumenti sifatida `list` obyektini berishimiz zarur:

In [2]:
a = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
print(a)

[ 1  2  3  4  5  6  7  8  9 10]


Yuqorida biz `list` obyektini NumPyning `ndarray` turiga o'tkazdik. Endi biz ushbu `ndarray`ning `a` obyekting xususiyatlari bilan tanishib chiqamiz. `a.shape` xususiyati `a` da nechta o'lcham bor hamda har bir o'lchamda qanchadan element borligini `tuple` obyekti sifatida qaytaradi:

In [3]:
print(a.shape)

(10,)


Demak `a` 1 o'lchamli tensor(vektor) hisoblanib, unda 10 ta element bor. Keling endi ikki o'lchamli tensor(matritsa)ni ham bir qaraylik:

In [49]:
A = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
    [10, 11, 12]
])
print(A.shape)

(4, 3)


**Yodda tuting.** Biz faqat to'g'iri to'rtburchak shakldagi tensorlarni yarata olamiz, ya'ni har bir o'qda bir xil sonda elementlar bo'lishi kerak. Masalan, quyidagicha obyekt yarata olmaymiz:

In [50]:
A = np.array([
    [1, 2, 3, 5],
    [4, 5, 6],
    [7, 8, 9, 6],
    [10, 11, 12]
])
print(A.shape)

ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (4,) + inhomogeneous part.

Yuqorida biz matritsaning birinchi qatorida 4 element ikkinchi qatorida esa 3 ta elementdan iborat bo'lishini xohlayabmiz, buning esa iloji yo'q. Shuning uchun ham har bir o'qda bir xil sondagi elementlari bo'lishi zarur. Bunda cheklov `list` turida yo'q, chunki u ichma ich obyektlarni saqlaydi.

Demak `A` obyekti 4 ta qator va 3 ta ustunda iborat matritsa(2 o'lchovli tensor) ekan. Biz ushbu jarayoni yana takrorlash orqali 3 o'lchovli tensor yoki n o'lchovli tensor ham yaratishimiz mumkin. Lekin biz ikki o'lchovlidan kattasini tassavur qila olmaymiz yoki biror qiymat bilan to'ldirishni tushuna olmaymiz. Biroq ilmiy hisoblarda esa biz ulardan doimiy foydalanamiz. Masalan, ramslarni RGB shaklda saqlasak, unda bizga bitta rasm uchun 3 o'lchovli tensor kerak bo'ladi. Agar rasmlar bittadan ko'p bo'lsa, 4 o'lchovlidan foydalanamiz. Agar bizga faqat tensorning necha o'lovchligi zarur bo'lsa, u holda `len` funksiyasidan foydalanishimiz mumkin:

In [10]:
print(len(A.shape))

2


Ba'zida umumiy emelentlar sonini aniqlashga to'g'iri keladi, bu holda esa `size` xususiyatidan foydalanish mumkin bo'ladi:

In [11]:
print(A.size)

12


### Obyektlarni yaratish funksiyalari

Ba'zi hollarda elementlari ma'lum bir son(lar)dan iborat bo'lgan bir jinsli tensor yaratish zarurati tug'iladi. Buning uchun biz NumPyning mavjud funkisyalaridan yodalanishimiz mumkin bo'ladi:

* `np.zeros()` -- qiymati noldan iborat bo'lgan tensor yaratib beradi:

In [61]:
a = np.zeros(shape=(10, ))
print("a=", a)

A = np.zeros(shape=(3, 4))
print("A=", A)

a= [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
A= [[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]


* `np.ones()` --  qiymati birdan iborat bo'lgan tensor hosil qiladi:

In [62]:
a = np.ones(shape=(10, ))
print("a=", a)

A = np.ones(shape=(3, 4))
print("A=", A)

a= [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
A= [[1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]]


Ushbu ikki funksiyada ham bir qancha argumentlar mavjud, ularda biri `dtype` hisoblandi. Bu agrument orqali biz hosil bo'layotgan obyektning turini oldindan belgilashimiz mumkin:

In [66]:
a = np.ones(shape=(10, ), dtype=np.int32)
print("a=", a)
print(a.dtype)

A = np.ones(shape=(3, 4), dtype=np.int32)
print("A=", A)
print(A.dtype)

a= [1 1 1 1 1 1 1 1 1 1]
int32
A= [[1 1 1 1]
 [1 1 1 1]
 [1 1 1 1]]
int32


Ushbu funksiyalarning quyidagi ko'rinishi ham mavjud: `np.zeros_like()` va `np.ones_like()`. Bu ikki funksiya berilgan tensorga o'xshash o'lchamdagi qiymatlari 0 yoki bir bo'lgan tensorlar hosil qiladi:

In [67]:
a = np.array([1, 2, 3, 4])
b = np.ones_like(a)
print("b=", b)

A = np.array([
    [1, 2, 3],
    [4, 5, 6]
])
B = np.ones_like(A)
print("B=", B)

b= [1 1 1 1]
B= [[1 1 1]
 [1 1 1]]


Yuqoridagi ikkita funskiyaga qo'shimcha ravishda bir qancha funksiayalarimizda kelgusi mavzularda batafsil o'rganib chiqamiz. 

## Qiymat(lar)ni olish va berish

Ushbu ikki amal NumPyning sal qiyinroq hamda, shu bilan birgalikda, yana bir muhim afzaligi hisoblanadi. Keling namunalarni matrtisalar ustida ko'rib chiqaylik, shunda biz matritsadagi o'zgarishlarni tushunishimiz oson bo'ladi. Bu ikki amalda ham bir vaqtda ham `list` obyektida bo'lgani kabi indekslash bor ham kesish amallari bor, faqat `list` turidan farqli ravishda biz hamma indekslarni bitta `[]` qavs ichiga o'qlarni vergul orqali ajratib yozamiz.

### Qiymat(lar)ni olish

Quyida birinchi `A` obyektniga matritsani yuklaylik hamda uning 1-qator hamda 2-ustun elementining qiymatini chop qilaylik.

> **Eslatma.** Pythondagi kabi Numpyning `ndarray` obyektida ham indekslash 0 sonidan boshlanadi.

In [41]:
A = np.array([
    [1, 2, 3, 4],
    [5, 6, 7, 8],
    [9, 10, 11, 12]
])
print(A[1, 2])

7


Agar biror o'qdagi to'liq qiymatlarni olmoqchi bo'lsa, u holda biz yana `ndarray` obyektiga ega bo'lamiz. Masalan, yuqoridagi matritsaning 1-chi qatorini olaylik.

In [42]:
print(A[1])

[5 6 7 8]


Oldingi misolda, qatorni olish oddiy, chunki matritsaning qatori `ndarray` obyektining birinchi o'qida joylashgan. Lekin biz 0-ustuni olmoqchi bo'lsak, u holda `:` kesish operatorini birinchi(qator) o'qqiga quyishimiz kerak hamda ikkinchi o'q(ustun)ning sonini esa verguldan keyin yozishimiz zarur bo'ladi:

In [43]:
print(A[:, 0])

[1 5 9]


Yuqoridagi misoldan sodda bitta narsani biz aniqlashimiz mumkin, bu `:` kesish operatori o'zi kelganda butun boshli o'qdagi hamma elementi oladi. Shunga asosan biz `list` obyektidagi kesish qoidalarini `ndarray` obyektlari uchun ham ishlatishimiz mumkin bo'ladi. Masalan, keling matritsaning 1-qatoridagi o'rtadagi ikki elementni (6, 7) olaylik. Buning uchun biz 1-qatorni to'liq tanlab, ustundan 1-dan 3-gacha bo'lganlarini tanlashimiz kerak, ya'ni `1:3`. Bunda 3 kirmaydi, shuning uchun ham bitta ortiqcha qiymat yozayabmiz.

> **Eslatma.** Agar biz matritsaning birinchi qatorining qiymatlarini olsak hamda ustuniga `:` kesish amalini qo'ysak(`print(A[1, :])`) ham birinchi bundan oldingi misoldagi natijani olamiz, faqat ortiqcha kod yoziladi xolos.

In [45]:
print(A[1, 1:3])

[6 7]


Agar yuqoridagi misolda biz birinchi o'qni to'liq tanlasak, u holda matritsaga ega bo'lamiz. Ushbu matritsa hamma qatorlardagi 1-2-elementlardan iborat bo'ladi.

In [46]:
print(A[:, 1:3])

[[ 2  3]
 [ 6  7]
 [10 11]]


Endi `list` obyektida mavjud bo'lgan `start:stop:step` amali, ya'ni kesish amalini to'liqroq ko'rib chiqamiz. Bunda biz har doim har bir o'qni alohida qaraymiz. Keling `A` matritsadgi 0 va 1-ustun qiymatlarini olaylik. Bu holda biz `step=2` deb olishimiz zarur bo'ladi. Kelishuv bo'yicha esa bu `step=1`.

In [48]:
print(A[:, ::2])

[[ 1  3]
 [ 5  7]
 [ 9 11]]


Yuqorida faqat `step` parametrga qiymat berdik, qolganlariga agar qiymat berilmasa, holda ular mos ravishda 0 va o'qdagi elementlar sonini oladi.

Oldin ta'kidlaganimizdek, NumPyning kengaytirilgan imkoniyatlari qismida ushbu bo'limga qo'shimchalar qo'shamiz.

### Qiymat(lar) berish

Qiymatlar berish ham olishga o'xshash amalga oshadi odatda, faqat qiymat berishda biz ba'zi cheklovlarga e'tibor berishimiz zarur bo'ladi:

* berilayotgan qiymat bilan ajratilgan sohaning o'lchami bir xil bo'lishi yoki umumiy qoidalarga mos tushishi zarur;
* 

`ndarray` obyketi o'zgaruvchan(muttable) hisoblangani uchun biz osongina qiymatlari indekslash amallari orqali berishimiz mumkin bo'ladi. Masalan, quyida matritsaning 2-qatorining 1-ustuni elementini o'zgartiraylik:

In [51]:
A = np.array([
    [1, 2, 3, 4],
    [5, 6, 7, 8],
    [9, 10, 11, 12]
])
A[2, 1] = 100
print(A)

[[  1   2   3   4]
 [  5   6   7   8]
 [  9 100  11  12]]


Endi biror o'qning qiymatini to'g'icha bitta qiymat bilan almashtiraylik:

In [52]:
A = np.array([
    [1, 2, 3, 4],
    [5, 6, 7, 8],
    [9, 10, 11, 12]
])
A[0] = 0
print(A)

[[ 0  0  0  0]
 [ 5  6  7  8]
 [ 9 10 11 12]]


Ko'rib turganimizdek biz butun qatorgai hamma elementlarni 0 qiymatiga o'zgartirdik. Agar biror ustun uchun shu amalni bajarmoqchi bo'lsak, u holda shunchaki `:` kesish operatoridan foydalanishimiz yetarli bo'ladi:

In [53]:
A = np.array([
    [1, 2, 3, 4],
    [5, 6, 7, 8],
    [9, 10, 11, 12]
])
A[:, 2] = 0
print(A)

[[ 1  2  0  4]
 [ 5  6  0  8]
 [ 9 10  0 12]]


Xuddi shu kabi biror sohaga ham qiymat bersak bo'ladi. Bu holda faqat ma'lum bir sohani tanlashimiz zarur bo'ladi. Sohani tanlashni esa kesish operatori orqali bajarishini yuqorida ko'rib chiqdik.

In [56]:
A = np.array([
    [1, 2, 3, 4],
    [5, 6, 7, 8],
    [9, 10, 11, 12]
])
A[1:, :2] = 0
print(A)

A = np.array([
    [1, 2, 3, 4],
    [5, 6, 7, 8],
    [9, 10, 11, 12]
])
A[1:, ::2] = 0
print(A)

[[ 1  2  3  4]
 [ 0  0  7  8]
 [ 0  0 11 12]]
[[ 1  2  3  4]
 [ 0  6  0  8]
 [ 0 10  0 12]]


Bunday misollarni ko'plab keltirish mumkin. Keling endi bitta son emas `ndarray` obyektini biror sohaga beraylik. Bunining uchun birinchi navbatda sohalarning o'lchami mos bo'lishi kerak.

In [57]:
A = np.array([
    [1, 2, 3, 4],
    [5, 6, 7, 8],
    [9, 10, 11, 12]
])
A[1:, ::2] = np.array([
    [11, 22],
    [33, 44]
])
print(A)

[[ 1  2  3  4]
 [11  6 22  8]
 [33 10 44 12]]


Qiymat berishining yana bir turi bu quyidagi teng kuchli misolarda ko'rishimiz mumkin. 

In [58]:
A = np.array([
    [1, 2, 3, 4],
    [5, 6, 7, 8],
    [9, 10, 11, 12]
])
A[1:, ::2] = np.array([[11, 22]])
print(A)

[[ 1  2  3  4]
 [11  6 22  8]
 [11 10 22 12]]


In [59]:
A = np.array([
    [1, 2, 3, 4],
    [5, 6, 7, 8],
    [9, 10, 11, 12]
])
A[1:, ::2] = np.array([11, 22])
print(A)

[[ 1  2  3  4]
 [11  6 22  8]
 [11 10 22 12]]


Yuqoridagi ikki misol ham bir xil vazifani bajarayabdi. Bu kabi yuqori darajadagi misollarni keyinchalik yana qayta o'rganib chiqamiz.

## Element bo'yicha amallar

Ushbu kutubxona yuqorida takidlaganimizdek juda ko'p imkoniyatlarga ega bo'lib, hozir shulardan birinchisi -- vektorlashgan amallarni ko'rib chiqamiz. Biz vektorlashgan amallar deganda `for` yoki `while` operatorlari orqali takrorlash asosida bajarildigan algorimtlarni biz shu operatorlarsiz tezlikda bajarishimiz mumkin bo'ladi. Keling soddalik uchun birinchi 10 000 000 sondan iborat bo'lgan vektorning har bir elementiga biror soni qo'shib ko'raylik. Avallo `list` va `ndarray` turida ikkita qiymatlari bir bo'lgan obyekt yaratamiz.

In [30]:
n = 10_000_000
py_nums = list(range(n))
numpy_nums = np.array(py_nums)

So'ng esa `list` obyektning har bir elementining qiymatini 1 ga oshirib chiqamiz, buning uchun biz `for` operatoridan quydagicha foydalanamiz:

In [31]:
%timeit for i in range(n): py_nums[i] += 1

593 ms ± 4.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


Yuqorida `%timeit` sehrli buyrug'i biz shu qator uchun ketgan vaqtni hisoblash hamda qolgan statistikalarni chiqarishga yordam beradi. Xususan, bu kod jami 7 marta ishga tushirilib, keyin natijalarning o'rtachasi olingan.

In [32]:
%timeit numpy_nums1 = numpy_nums + 1

15.6 ms ± 137 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


Ko'rib turganimizdek, oddiy `list` obyekti uchun 593 ms vaqt ketgan bo'lsa, `ndarray` obyekti uchun esa 15.6 ms safrlangan. Bundan tashqari kodimiz ham ancha sodda yozilgan Numpy asosida.

Numpy tensorlari ustida amalar bajargan biz bir qancha narsalarga sinchikovlik bilan e'tibor berishimiz zarur, aks holda biz o'zimiz o'ylamagan natijaga erishamiz yoki murakkab masalalarni takrorlash operatorlarisiz bajara olmay qolamiz. Masalan, yuqorida biz `numpy_nums` obyektning har bir elementini bittaga ochiqib `numpy_nums1` obyektini hosil qildik, lekin takrorlash amallaridan foydalanmadik. Bu kod uchun biz qisqacha qilib quyidagi savolni berishimiz mumkin "Aslida vektorga soni qo'shish mumkinmi (matematik jihatdan)?" Ko'pchilik aytishi mumkin "yo'q deyishi aniq." Lekin NumPy kutubxonasi ilmiy masalalarni yechishda tensorlar ustida amallarni imkon qadar optimlashtirish orqali amalga oshirgan va buni biz ingliz tilida "broadcasting" deb ataymiz. Shuning uchun ham yuqoridagi kodda ikkita turdagi obyekt o'rtasida qo'shish amali bajarilyabdi va natija esa "broadcasting" qoidalari asosida. Bu haqida rasmiy sahifada batafsil bilib olishingiz mumkin: **[NumPy broadcasting](https://numpy.org/doc/stable/user/basics.broadcasting.html)**.

### NumPy obyekti va son ustida amallar

To'rta arfimetik amallar hamda `%`, `//` amallari tensorning o'lchami qanday bo'lishidan qattiy nazar tensorning har bir elementiga alohida qo'laniladi. Masalan, quyidagi matritsani va 2 sonini qaraylik.

In [33]:
A = np.array([
    [1, 2, 3],
    [4, 5, 6],
])
print(A)

[[1 2 3]
 [4 5 6]]


In [38]:
B = A + 2
print("A+2=\n", B)
B = A - 2
print("A-2=\n", B)
B = A * 2
print("A*2=\n", B)
B = A / 2
print("A/2=\n", B)
B = A // 2
print("A//2=\n", B)
B = A % 2
print("A%2=\n", B)

A+2=
 [[3 4 5]
 [6 7 8]]
A-2=
 [[-1  0  1]
 [ 2  3  4]]
A*2=
 [[ 2  4  6]
 [ 8 10 12]]
A/2=
 [[0.5 1.  1.5]
 [2.  2.5 3. ]]
A//2=
 [[0 1 1]
 [2 2 3]]
A%2=
 [[1 0 1]
 [0 1 0]]


Yuqoridagi natijalardan ma'lumki 2 sonini `A` tensorining har bir elementiga o'ng tomondan amalga mos ta'sir qilmoqda. Keyingi holat esa agar shu son amalning chap tomonida kelsa, qanday natija bo'ladi shuni ko'raylik:

In [39]:
B = 2 + A
print("A+2=\n", B)
B = 2 - A
print("A-2=\n", B)
B = 2 * A
print("A*2=\n", B)
B = 2 / A
print("A/2=\n", B)
B = 2 // A
print("A//2=\n", B)
B = 2 % A
print("A%2=\n", B)

A+2=
 [[3 4 5]
 [6 7 8]]
A-2=
 [[ 1  0 -1]
 [-2 -3 -4]]
A*2=
 [[ 2  4  6]
 [ 8 10 12]]
A/2=
 [[2.         1.         0.66666667]
 [0.5        0.4        0.33333333]]
A//2=
 [[2 1 0]
 [0 0 0]]
A%2=
 [[0 0 2]
 [2 2 2]]


Ushbu holatda ham natija tensor bo'ladi, faqat 2 sonini `A` tensorning har elementiga ko'rsatilgan amal asosida natija hosil bo'ladi. Xuddi shunday tarizda, boshqa ixtiyoriy o'lchamdagi tensor uchun yuqoridagidek holatni qo'llashimiz mumkin bo'ladi.

### Bir xil o'lchamli tensorlar ustida amallar

Agar tensorlarning ikkisining o'lchamlari bir xil bo'lsa, u holda mos elementlarining o'rtasidagi qaralayotgan amal oddiy ikkita son ustida bo'lgani kabi ishlaydi. Quyida shunga misollar ko'rishimiz mumkin.